File: typescript/src/verification/terminal-places.ts
/**
 * @module terminal-places
 *
 * The verification encoding of net-declared terminal places ([EXEC-042], [VER-014]).
 *
 * A terminal place `P` ends the run the moment a deposit marks it. The verifiers see exactly that
 * behaviour in a plain net where `P` inhibits every transition — no transition fires once `P` is
 * marked — with `P` a sink ([VER-002]) and a conditional-sink marker for every place
 * (`sinkPlacesWhen(P, all places)`): a marking with `P` marked is quiescent and excused.
 *
 * The caller restates none of this. A net without terminals is never touched, so its scripts
 * stay byte-identical ([VER-013]).
 */
import { PetriNet } from '../core/petri-net.js';
import type { Place } from '../core/place.js';
import { Transition } from '../core/transition.js';

/**
 * `net` with each of its terminal places inhibiting every transition, or `net` itself — the
 * same instance — when it declares no terminals, so its scripts stay byte-identical.
 *
 * Names, place order and transition order are kept; the added inhibitors follow each
 * transition's own, in terminal declaration order, and one already present is not repeated.
 * The result declares **no** terminals — the inhibitors now carry them — so applying this twice
 * is the same as applying it once. The caller adds the sink half of the encoding (each terminal
 * a sink and a conditional-sink marker for {@link terminalExcusedPlaces}), which a net cannot
 * say by itself. Mirrors Java's `TerminalEncoding.inhibited`.
 */
export function withTerminalInhibitors(net: PetriNet): PetriNet {
  if (net.terminals.size === 0) return net;
  const terminals = [...net.terminals];
  const transitions = [...net.transitions].map(t => {
    const inhibited = new Set(t.inhibitors.map(arc => arc.place.name));
    const missing = terminals.filter(p => !inhibited.has(p.name));
    return missing.length === 0 ? t : withInhibitors(t, missing);
  });
  return PetriNet.builder(net.name)
    .places(...net.places)
    .transitions(...transitions)
    .build();
}

/**
 * The places each terminal excuses while it is marked: every place of `net`, in net order.
 * `sinkPlacesWhen(P, …these)` and the open-net `DesignedTerminal` both take this list.
 */
export function terminalExcusedPlaces(net: PetriNet): Place<any>[] {
  return [...net.places];
}

/** `t` with inhibitor arcs from `places` appended after its own; everything else is carried. */
function withInhibitors(t: Transition, places: readonly Place<any>[]): Transition {
  const b = Transition.builder(t.name)
    .timing(t.timing)
    .priority(t.priority)
    .action(t.action);
  if (t.placeAlias.size > 0) b.placeAlias(t.placeAlias);
  if (t.inputSpecs.length > 0) b.inputs(...t.inputSpecs);
  if (t.outputSpec !== null) b.outputs(t.outputSpec);
  for (const arc of t.inhibitors) b.inhibitor(arc.place);
  for (const p of places) b.inhibitor(p);
  for (const arc of t.reads) b.read(arc.place);
  for (const arc of t.resets) b.reset(arc.place);
  if (t.matchSpec !== null) b.match(t.matchSpec);
  return b.build();
}

File: rust/libpetri-verification/src/terminal_places.rs
//! Net-declared terminal places in verification ([EXEC-042], [VER-014]).
//!
//! A terminal place `P` ends the run the moment it is marked: nothing fires
//! afterwards, whatever else is enabled or in flight. The verifier states the
//! same thing in the model, with no restatement by the caller, by verifying
//! the net in which
//!
//! - `P` inhibits every transition (so no firing follows a marking of `P`),
//! - `P` is a sink ([VER-002]), and
//! - `P` is a conditional-sink marker for every place ([VER-014]),
//!   `sink_places_when(P, all places)`,
//!
//! so a quiescent marking with `P` marked is excused. This is exact for the
//! runtime's strict stop. The inhibitor half lives here; each verifier adds the
//! rest-set half to its own declarations.

use libpetri_core::arc::Inhibitor;
use libpetri_core::petri_net::PetriNet;

/// The net with every declared terminal place inhibiting every transition, or
/// `None` when the net declares no terminal place — the caller then keeps the
/// original net, so a net without terminals is verified, and scripted, exactly
/// as before ([EXEC-042] AC8).
///
/// Inhibitors are appended after each transition's own, in terminal
/// declaration order, skipping a place the transition already inhibits, so the
/// rewrite is idempotent. The result still declares its terminals.
pub fn inhibit_on_terminals(net: &PetriNet) -> Option<PetriNet> {
    if net.terminals().is_empty() {
        return None;
    }
    let inhibitors: Vec<Inhibitor> = net
        .terminals()
        .iter()
        .map(|p| Inhibitor { place: p.clone() })
        .collect();
    Some(net.map_transitions(|t| t.with_added_inhibitors(inhibitors.iter().cloned())))
}

/// Every place name of `net`, in net order: the places a terminal excuses
/// ([VER-014] `sink_places_when(P, all places)`).
pub fn all_place_names(net: &PetriNet) -> Vec<String> {
    net.places().iter().map(|p| p.name().to_string()).collect()
}
